File: host/src/main/java/dk/dtu/domain/rules/effects/Pits.java
package dk.dtu.domain.rules.effects;

import dk.dtu.domain.core.Phase;
import dk.dtu.domain.model.Direction;
import dk.dtu.domain.model.Robot;
import dk.dtu.domain.model.Tile;
import dk.dtu.domain.rules.Coord;
import dk.dtu.domain.rules.DestroyCause;
import dk.dtu.domain.rules.api.BoardAPI;

import java.util.EnumSet;
import java.util.List;

/**
 * A pits tile kills any {@link Robot} standing on it
 * @author Weihao Mo
 */
public record Pits() implements TileEffect{
    /**
     * @author Weihao Mo
     */
    @Override
    public void onPhase(Phase phase, Tile tile, BoardAPI api) {
        int x = tile.getX();
        int y = tile.getY();
        for(Robot robot: api.getRobotsOnTile(x, y)) {
            api.reportDestroy(robot.getId(), new Coord(x, y), DestroyCause.PITS);
        }
    }

    /**
     * @author Weihao Mo
     */
    public static boolean hasPits(Tile t) {
        if (t == null) {
            return false;
        }
        for (TileEffect effect: t.getEffects()) {
            if (effect instanceof Pits) {
                return true;
            }
        }
        return false;
    }

    /**
     * @author Weihao Mo
     */
    @Override
    public EnumSet<Phase> phases() {
        return EnumSet.of(Phase.ACTIVATE_PITS);
    }
}

File: engine/src/effects/pits.ts
import { Phase } from "../core/phase.js";
import type { Tile } from "../model/tile.js";
import { Coord } from "../rules/coord.js";
import { DestroyCause } from "../rules/outcome.js";
import type { BoardAPI } from "../rules/boardApi.js";
import type { TileEffect } from "./tileEffect.js";

/** Ported from dk.dtu.domain.rules.effects.Pits. */
export class Pits implements TileEffect {
  onPhase(_phase: Phase, tile: Tile, api: BoardAPI): void {
    const x = tile.getX();
    const y = tile.getY();
    for (const robot of api.getRobotsOnTile(x, y)) {
      api.reportDestroy(robot.getId(), new Coord(x, y), DestroyCause.PITS);
    }
  }

  static hasPits(tile: Tile | null): boolean {
    if (tile === null) return false;
    for (const effect of tile.getEffects()) {
      if (effect instanceof Pits) return true;
    }
    return false;
  }

  phases(): Set<Phase> | null {
    return new Set([Phase.ACTIVATE_PITS]);
  }
}
